File: Data Structures & Algorithms/max-water-container/submission-5.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        
        L, R = 0, len(heights)-1
        
        res = 0

        while L < R:
            # update the result with the area (height of lowest bar)
            res = max(min(heights[L],heights[R]) * abs(R-L),res)

            # progress from lower bar inwards
            
            #if the right bar is higher move the left bar inwards
            if heights[R]>=heights[L]:
                L+=1
            # vice versa
            else:
                R-=1
            
        return res

File: Data Structures & Algorithms/house-robber/submission-5.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        vector<int> dp;
        int n = nums.size();
        if (n==1) return nums[0];
        dp.reserve(n);
        for (int i =0; i < n; i++) {
            if (i == 0) {
                dp[i] = nums[i];
            } else if (i==1) {
                dp[i] = max(nums[i],dp[i-1]);
            }
            else {
                dp[i] = max(nums[i] + dp[i-2],dp[i-1]);
            }
        }
        return max(dp[n-1],dp[n-2]);
    }
};
